### PR TITLE
Add `_name` support for top level `knn` clauses

### PR DESCRIPTION
**Pull request by Rassyan**
Resolves #106254

This PR introduces `_name` support for top-level kNN queries, enabling named query functionality consistent with other query types. Changes include serialization/deserialization of the `_name` field within `KnnSearchBuilder` and handling within `KnnScoreDocQueryBuilder`.

Key Changes:
- Added `_name` field to `KnnSearchBuilder`.
- Modified serialization to include `_name`.
- Ensured `_name` is processed during query execution and included in the response.

Tests:
- Updated existing tests to cover `_name` functionality.
- Added new tests to ensure correct serialization/deserialization and response behavior.

Please review the changes and let me know if any further modifications are required.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java [hunk 1]
```diff
@@ -459,6 +485,7 @@ public void writeTo(StreamOutput out) throws IOException {
         }
         out.writeNamedWriteableCollection(filterQueries);
         out.writeFloat(boost);
+        out.writeOptionalString(queryName);
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java:488]: The reading and writing to the transport stream needs to be protected and backwards compatible. 

There are examples of this elsewhere, but a new version needs added to Transport versions and you should check the stream version before reading and writing.

File: server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java [hunk 2]
```diff
@@ -372,7 +392,8 @@ public KnnVectorQueryBuilder toQueryBuilder() {
         if (queryVectorBuilder != null) {
             throw new IllegalArgumentException("missing rewrite");
         }
-        return new KnnVectorQueryBuilder(field, queryVector, numCands, similarity).boost(boost).addFilterQueries(filterQueries);
+        return new KnnVectorQueryBuilder(field, queryVector, numCands, similarity)
+            .boost(boost).queryName(queryName).addFilterQueries(filterQueries);
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java:396]: Ah! Nice it's so simple here!

File: server/src/main/java/org/elasticsearch/action/search/DfsQueryPhase.java [hunk 3]
```diff
@@ -155,7 +155,7 @@ ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request) {
                 scoreDocs.toArray(new ScoreDoc[0]),
                 source.knnSearch().get(i).getField(),
                 source.knnSearch().get(i).getQueryVector()
-            ).boost(source.knnSearch().get(i).boost());
+            ).boost(source.knnSearch().get(i).boost()).queryName(source.knnSearch().get(i).queryName());
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/action/search/DfsQueryPhase.java:158]: ❤️

**Review by benwtrent — COMMENTED** [review]
This is a good start!

There does need to be backwards compatibility protection for mixed cluster versions.

It would also be good to have some 'yaml' tests. There are various "knn_search.yml" files. One of the basic ones should be updated to have one that adds a test that includes the name for knn and asserts it's returned.

The yaml tests add a fair bit of coverage as they are used in mixed cluster environments, cross cluster search, security, etc.

File: server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java (hunk 1 above)
Comment (Rassyan) [reply server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java]: Just pushed a commit that adds version checks for the transport stream and updates the Transport versions. Commit hash is [[7454e8b](https://github.com/elastic/elasticsearch/pull/107645/commits/7454e8b166baec952e8e2a6829d73c3546bdf1f7)]. Could you take another look and let me know if it's all good now?

**Comment by benwtrent** [comment]
buildkite test this please

**Comment by benwtrent** [comment]
buildkite test this please

**Comment by benwtrent** [comment]
buildkite test this please

**Comment by benwtrent** [comment]
buildkite test this please

**Review by benwtrent — APPROVED** [review]
Great stuff @Rassyan ! I will merge as soon as CI is ✅ again :)

**Comment by benwtrent** [comment]
buildkite test this please

**Comment by benwtrent** [comment]
🎉 🎉 🎉 Thank you @Rassyan for your contribution 🎉 🎉 🎉

**Pull request merged**